tfhe/src/high_level_api/shortints/tests.rs

```diff
@@ -169,11 +169,11 @@ fn test_programmable_bootstrap_fhe_uint2() -> Result<(), Box<dyn std::error::Err
 
     let mut a = FheUint2::try_encrypt(2, &keys)?;
 
-    let c = a.map(|value| value * value);
+    let c = a.map(|value| (value * value) % 4);
     let decrypted = c.decrypt(&keys);
     assert_eq!(decrypted, (2 * 2) % 2);
 
-    a.apply(|value| value * value);
+    a.apply(|value| (value * value) % 4);
```

Comment:
This should not be needed here, the modulus should be applied internally, not at the user code level

Reply:
You mean `map` and `apply` should apply the modulus on the output of the closure itself?

If so, would you do the modulus in `GenericShortIntServerKey::bootstrap_(inplace)_with`?

Should the modulo be over the message space, the message and the carry space or the message space, carry space and padding bit? It seems to me that it depends on the user. We could have a function for each or leave it to the caller as it is now

Reply:
To me in the HLP API, we expose things like FheUint2, FheUint3 where the modulus is 2**2, 2**3
the idea is that the carry system is hidden, so users only see the 2/3 bits of message

Reply:
Makes sense, will fix

Reply:
Done

Reply:
Also the modulus used in the assert is wrong : it does `%2` but it should be "%4" 

(The 'bug' was there before your modifications)